### PR TITLE
Blazor Server SignalR guidance

### LINKED ISSUE #12949: Document enabling sticky sessions when using Blazor server apps with Azure SignalR service
[EDIT by guardrex to coalesce content from #15392 into this issue]
[EDIT by guardrex to add the topic metadata]
[EDIT by guardrex to remind myself to add the no-loc metadata: `no-loc: [Blazor]`]

Typically server-side Blazor apps will prerender in response to the first request which will setup the UI state on that server. When the client tries to then setup the SignalR connection for that app and is redirected through the SignalR service, we need the client to end back up on the same server. There is an option in the Azure SignalR SDK to support this scenario by enabling sticky sessions. You can turn it on by setting the "Azure:SignalR:StickyServerMode" config value to "Required".

[EDIT ... the following is from #15392 ...]

* Include a sample on how to configure it on Kubernetes as per https://github.com/aspnet/AspNetCore/issues/16645#issuecomment-547962457
* A sample on how to configure in App Service
This is done in the portal in Application Settings
* A sample on how to configure it in IIS
https://docs.microsoft.com/en-us/iis/extensions/configuring-application-request-routing-arr/http-load-balancing-using-application-request-routing

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 542f38c4-1502-f739-372a-80c86c377491
* Version Independent ID: 72c2a844-ffb2-4799-2c7a-a7a73d3b6833
* Content: [ASP.NET Core Blazor hosting models](https://docs.microsoft.com/en-us/aspnet/core/blazor/hosting-models?view=aspnetcore-3.0)
* Content Source: [aspnetcore/blazor/hosting-models.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/hosting-models.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### LINKED ISSUE #15552: Update the docs about reconnection to reflect
https://github.com/aspnet/AspNetCore/issues/16870#issuecomment-550345922

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 542f38c4-1502-f739-372a-80c86c377491
* Version Independent ID: 72c2a844-ffb2-4799-2c7a-a7a73d3b6833
* Content: [ASP.NET Core Blazor hosting models](https://docs.microsoft.com/en-us/aspnet/core/blazor/hosting-models?view=aspnetcore-3.0)
* Content Source: [aspnetcore/blazor/hosting-models.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/hosting-models.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #15552
Fixes #12949

* [Hosting Models internal review topic (*Connection to the server* section)](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/hosting-models?view=aspnetcore-3.0&branch=pr-en-us-15638#connection-to-the-server)
* [Blazor Server Host and Deploy internal review topic (*SignalR configuration* section)](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/blazor/server?view=aspnetcore-3.0&branch=pr-en-us-15638#signalr-configuration)

Notes:

* Reconnection model updates
* Sticky sessions configuration
* I **do not** address the remark in the SignalR *scale.md* topic ...

  > Sticky sessions, also known as [client affinity](/iis/extensions/configuring-application-request-routing-arr/http-load-balancing-using-application-request-routing#step-3---configure-client-affinity), is not required, because clients are immediately redirected to the Azure SignalR Service when they connect.

  ... which seems incorrect without the configuration that we're describing here. Let me know if/how you would like that updated, and I can roll the update in on this PR. It's in the live topic here ...

  https://docs.microsoft.com/en-us/aspnet/core/signalr/scale?view=aspnetcore-3.0#azure-signalr-service

* Note that https://github.com/Azure/azure-signalr/blob/dev/docs/use-signalr-service.md#serverstickymode doesn't seem to be covered in SignalR docs. I recommend a SignalR doc author take that on tho if coverage is desired ... I don't have much time left here with lot of work to do before the holidays 🎁🎅🦌. Let me know if you want an issue opened and which topic?

* 🚨 **Delta between issue remark and code** 🚨: The issue says that the config key is named `Azure:SignalR:StickyServerMode`, but the option is `options.ServerStickyMode` ... was it just a typo in the issue remark? If it's just a typo, no worries on this PR ... I used `Azure:SignalR:ServerStickyMode` here.

### CODE REVIEW

File: aspnetcore/host-and-deploy/blazor/server.md [hunk 1]
```diff
@@ -46,11 +46,50 @@ Blazor Server apps use ASP.NET Core SignalR to communicate with the browser. [Si
...
+1. Add the **Azure SignalR Service** dependency to the profile. If the Azure subscription doesn't have a pre-existing Azure SignalR Service instance to assign to the app, select **Create a new Azure SignalR Service instance** to provision a new service instance.
+1. Publish the app to Azure.
+
+#### IIS
+
+When using IIS, sticky sessions are enabled with Application Request Routing. For more information, see [HTTP Load Balancing using Application Request Routing](/iis/extensions/configuring-application-request-routing-arr/http-load-balancing-using-application-request-routing).
+
+#### Kubernetes
+
+Create an ingress definition with the following settings:
```
Comment (scottaddie) [inline aspnetcore/host-and-deploy/blazor/server.md]:
```suggestion
Create an ingress definition with the following Kubernetes annotations:
```

File: aspnetcore/host-and-deploy/blazor/server.md [hunk 2]
```diff
@@ -46,11 +46,50 @@ Blazor Server apps use ASP.NET Core SignalR to communicate with the browser. [Si
...
+
+```
```
Comment (scottaddie) [inline aspnetcore/host-and-deploy/blazor/server.md]: I recommend adding a brief explanation of what these annotations do. You can find that info at https://github.com/kubernetes/ingress-nginx/blob/master/docs/user-guide/nginx-configuration/annotations.md#session-affinity.

File: aspnetcore/blazor/hosting-models.md [hunk 3]
```diff
@@ -123,18 +124,38 @@ Blazor Server apps should be optimized to minimize UI latency by reducing networ
...
+    ...
+</div>
+```
+
+The following table describes the CSS classes applied to the `components-reconnect-modal` element.
+
+| CSS class                       | Indicates&hellip; |
+| ------------------------------- | ------------------------- |
+| `components-reconnect-show`     | A lost connection. The client is attempting to reconnect. Show the modal. |
+| `components-reconnect-hide`     | An active connection is re-established to the server. Hide the model. |
```
Comment (scottaddie) [inline aspnetcore/blazor/hosting-models.md]:
```suggestion
| `components-reconnect-hide`     | An active connection is re-established to the server. Hide the modal. |
```

File: aspnetcore/blazor/hosting-models.md [hunk 4]
```diff
@@ -123,18 +124,38 @@ Blazor Server apps should be optimized to minimize UI latency by reducing networ
...
-* `components-reconnect-hide` &ndash; The client has an active connection, hide the UI.
-* `components-reconnect-failed` &ndash; Reconnection failed, probably due to a network failure. To attempt reconnection, call `window.Blazor.reconnect()`.
-* `components-reconnect-rejected` &ndash; Reconnection rejected. The server was reached but refused the connection, and the user's state on the server is gone. To reload the app, call `location.reload()`. This connection state may result when:
-  * A crash in the circuit (server-side code) occurs.
-  * The client is disconnected long enough for the server to drop the user's state. Instances of components that the user was interacting with are disposed.
+A Blazor Server app prerenders in response to the first client request, which sets up the UI state on the server. When the client attempts to create a SignalR connection and is redirected through the SignalR service, the client must reconnect to the same server. Blazor Server apps that use more than one backend SignalR server should implement *sticky sessions* for SignalR connections.
```
Comment (danroth27) [inline aspnetcore/blazor/hosting-models.md]:
```suggestion
A Blazor Server app prerenders in response to the first client request, which sets up the UI state on the server. When the client attempts to create a SignalR connection, the client must reconnect to the same server. Blazor Server apps that use more than one backend server should implement *sticky sessions* for SignalR connections.
```

File: aspnetcore/blazor/hosting-models.md [hunk 5]
```diff
@@ -123,18 +124,38 @@ Blazor Server apps should be optimized to minimize UI latency by reducing networ
...
+```
+
+The following table describes the CSS classes applied to the `components-reconnect-modal` element.
+
+| CSS class                       | Indicates&hellip; |
+| ------------------------------- | ------------------------- |
+| `components-reconnect-show`     | A lost connection. The client is attempting to reconnect. Show the modal. |
+| `components-reconnect-hide`     | An active connection is re-established to the server. Hide the modal. |
+| `components-reconnect-failed`   | Reconnection failed, probably due to a network failure. To attempt reconnection, call `window.Blazor.reconnect()`. |
+| `components-reconnect-rejected` | Reconnection rejected. The server was reached but refused the connection, and the user's state on the server is lost. To reload the app, call `location.reload()`. This connection state may result when:<ul><li>A crash in the server-side circuit occurs.</li><li>The client is disconnected long enough for the server to drop the user's state. Instances of the components that the user is interacting with are disposed.</li></ul> |
```
Comment (danroth27) [inline aspnetcore/blazor/hosting-models.md]: This also often happens when the server gets restarted or recycled.

**Pull request merged**